tests/topotests/multicast_pim6_dual_stack/test_multicast_pim6_dual_stack.py

```diff
@@ -0,0 +1,1936 @@
+    }
+    result = create_mld_config(tgen, topo, input_dict_1)
+    assert result is True, "Testcase {}: Failed Error: {}".format(tc_name, result)
+
+    step("Add IGMP config from receiver pots")
+    input_dict_1 = {
+        "r1": {"igmp": {"interfaces": {intf_r1_i1: {"igmp": {"version": "2"}}}}}
+    }
+    result = create_igmp_config(tgen, topo, input_dict_1)
+    assert result is True, "Testcase {}: Failed Error: {}".format(tc_name, result)
+
+    for data in input_dict_v6:
+        result = verify_mroutes(
+            tgen,
+            data["dut"],
+            data["src_address"],
+            MLD_JOIN_RANGE_1,
+            data["iif"],
+            data["oil"],
+        )
+        assert result is True, "Testcase {} : Failed Error: {}".format(tc_name, result)
+
+        result = verify_upstream_iif(
+            tgen, data["dut"], data["iif"], data["src_address"], MLD_JOIN_RANGE_1
+        )
+        assert result is True, "Testcase {} : Failed Error: {}".format(tc_name, result)
+
+    for data in input_dict_v4:
+        result = verify_mroutes(
+            tgen,
+            data["dut"],
+            data["src_address"],
+            IGMP_JOIN_RANGE_1,
+            data["iif"],
+            data["oil"],
+        )
+        assert result is True, "Testcase {} : Failed Error: {}".format(tc_name, result)
+
+        result = verify_upstream_iif(
+            tgen, data["dut"], data["iif"], data["src_address"], IGMP_JOIN_RANGE_1
+        )
+        assert result is True, "Testcase {} : Failed Error: {}".format(tc_name, result)
+
+    r4_iif = (topo["routers"]["r4"]["links"]["i7"]["interface"],)
+    r4_oil = topo["routers"]["r4"]["links"]["r2"]["interface"]
+
+    step("From source node verify OIL towards RP is removed")
+    result = verify_mroutes(
+        tgen, "r1", source_v6, MLD_JOIN_RANGE_1, r4_iif, r4_oil, expected=False
+    )
+    assert result is not True, "Testcase {} : Failed Error: {}".format(tc_name, result)
+
+    result = verify_mroutes(
+        tgen, "r1", source_v6, IGMP_JOIN_RANGE_1, r4_iif, r4_oil, expected=False
+    )
+    assert result is not True, "Testcase {} : Failed Error: {}".format(tc_name, result)
+
+    step("verify ipv6 multicast traffic on R1 for all the grps")
+    result = verify_sg_traffic(tgen, "r1", MLD_JOIN_RANGE_1, source_v6, "ipv6")
+    assert result is True, "Testcase {} : Failed Error: {}".format(tc_name, result)
+
+    step("verify ipv6 multicast traffic on R1 for all the grps")
+    result = verify_sg_traffic(tgen, "r1", IGMP_JOIN_RANGE_1, source_v4, "ipv4")
+    assert result is True, "Testcase {} : Failed Error: {}".format(tc_name, result)
+
+    write_test_footer(tc_name)
+
+
+def test_ipv4_ipv6_mroutes_upstream_when_mld_and_igmp_join_sent_from_same_interface_p0(
+    request, app_helper
+):
+    """
+    Verify IPv4 and IPv6 mroute and upstream when mld and MLD join sent from same interface
+    of FRR1 and source present on same interface in FRR4
+    """
+
+    tgen = get_topogen()
```

Comment:
around 15 lines following this code is common in all the apis here, is it possible to move it into a common api? Just a suggestion if it make sense.

Reply:
It is a test standard we are following, if need to change then it will applicable everywhere. We can keep it as of now.